Add machine learning vs pysteps guidelines (#160)
pySTEPS/pySTEPS: doc/source/user_guide/machine_learning_pysteps.rst

.. _machine_learning_pysteps:

> Machine learning vs pysteps

Reviewer: not quite convinced by the title "Machine learning vs pysteps" for two main reasons:

* it sounds a bit like a competition between a python package and the whole machine learning ...., while here is more a matter of benchmarking machine learning with traditional persistence based approaches available in pysteps.
* in principle, pysteps represents a general framework/community for precipitation nowcasting, and as such we could imagine to implement ML-based models, too. 

For these reasons, I would propose to change the title into something a bit different, as for example "Benchmarking machine learning models with pysteps". 

What do you think?

===========================

> How to correctly compare the accuracy of machine learning against pysteps?

Reviewer: ```suggestion
How to correctly compare the accuracy of machine learning against traditional nowcasting methods available in pysteps?
```
or something similar, but you get my point ;-)

> Before starting, you need to ask yourself these two questions:

Reviewer: Starting what? Is this referring to the comparison with extrapolation or to the actual training of the ML model?

#. Do you only want to minimize prediction errors?
#. Do you also want to represent the prediction uncertainty? 

To achieve objective 1, it is sufficient to produce a single deterministic nowcast that filters out the unpredictable small-scale precipitation features.
However, this will create a nowcast that will become increasingly smooth over time.

To achieve objective 2, you need to produce a probabilistic or an ensemble nowcast (several ensemble members and realizations).

In weather forecasting (and nowcasting) we usually want to achieve both goals because it is impossible to predict the evolution of a chaotic system with 100% accuracy, especially space-time precipitation fields and thunderstorms! 

Machine learning and pysteps offer several methods to produce both deterministic and probabilistic nowcasts. 

> Therefore, if you want to compare machine learning-based nowcasts to pysteps you need to select the right method.

Reviewer: ```suggestion
Therefore, if you want to compare machine learning-based nowcasts to simpler extrapolation-based models, you need to select the right method.
```

> 1. Deterministic machine learning vs pysteps

Reviewer: ```suggestion
1. Deterministic nowcasting
```

--------------------------------------------

Deterministic nowcasts can be divided into:

a. Variance-preserving nowcasts, such as extrapolation nowcasts by Eulerian and Lagrangian persistence.
b. Error-minimization nowcasts, such as machine learning-based or ensemble mean nowcasts.

**Very important**: these two types of deterministic nowcasts are not directly comparable because they have a different variance! 
This is best explained by the decomposition of the mean squared error (MSE):

:math:`MSE = bias^2 + Var`

All deterministic machine learning algorithms that minimize the MSE (or a variation) will also inevitably minimize the variance of nowcast fields.
This is a natural attempt to filter out the unpredictable precipitation features, which would otherwise increase the variance (and the MSE).
The same principle holds for convolutional and/or deep neural network architectures, which also produce smooth nowcasts.

> Therefore, you should never directly compare an error-minimization machine learning nowcast to a variance-preserving radar extrapolation, as produced by the module pysteps.nowcasts.extrapolation. Instead, you should use the pysteps ensemble mean.

Reviewer: ```suggestion
Therefore, it is better to avoid directly comparing an error-minimization machine learning nowcast to a variance-preserving radar extrapolation, as produced by the module pysteps.nowcasts.extrapolation. Instead, you should use the pysteps ensemble mean.
```

Reviewer: Also, do you think it could be better to link the modules you list here and below directly to the corresponding documentation page?

> A pysteps "ensemble mean" can be generated using the modules pysteps.nowcasts.sprog, pysteps.nowcasts.sseps or pysteps.nowcasts.anvil. 

Reviewer: ```suggestion
A deterministic equivalent of the ensemble mean can be approximated using the modules pysteps.nowcasts.sprog or pysteps.nowcasts.anvil. 
```
I wouldn't include sseps here: it's a stochastic method and it is more computationally demanding than steps...

Another possibility, but more computationally demanding, is to average many ensemble members generated by the modules pysteps.nowcasts.steps or pysteps.nowcasts.sseps. 

Still, even by using the pysteps ensemble mean, it is not given that its variance will be the same as the one of machine learning predictions. 
Possible solutions are to:

> #. use a normalized MSE or other score,

Reviewer: ```suggestion
#. use a normalized MSE (NMSE) or another score accounting for differences in the variance between prediction and observation.
```

Reviewer: maybe also include the use of categorical scores based on high-intensity thresholds?

#. decompose the field with a Fourier (or wavelet) transform to compare features at the same spatial scales.

A good deterministic comparison of a deep convolutional machine learning neural network nowcast and pysteps is given in :cite:`FNPC2020`.

> 2. Probabilistic machine learning vs pysteps

Reviewer: ```suggestion
2. Probabilistic nowcasting
```

--------------------------------------------

Probabilistic machine learning regression methods can be roughly categorized into:

a. Quantile-based methods, such as quantile regression, quantile random forests and quantile neural networks.

> b. Ensemble-based methods, such as generative adversarial neural networks (GANNs) and variational auto-encoders (VAEs).

Reviewer: ```suggestion
b. Ensemble-based methods, such as generative adversarial networks (GANs) and variational auto-encoders (VAEs).
```

Quantile-based machine learning nowcasts are interesting, but can only estimate the probability of exceedance at a given point (see e.g. :cite:`FSNBG2019`).

To estimate areal exceedance probabilities, for example above catchments, or to propagate the nowcast uncertainty into hydrological models, the full ensemble still needs to be generated, e.g. with generative machine learning models.

Generative machine learning methods are similar to the pysteps ensemble members. Both are designed to produce an ensemble of possible realizations that preserve the variance of observed radar fields.

A proper probabilistic verification of generative machine learning models against pysteps would be an interesting research direction.

Summary
-------
The table below is an attempt to classify machine learning and pysteps nowcasting methods according to the four main prediction types:

#. Deterministic (variance-preserving)
#. Deterministic (error-minimization)
#. Probabilistic (quantile-based)
#. Probabilistic (ensemble-based)

The comparison of methods from different types should only be done carefully and with good reasons.

> .. list-table::

Reviewer: this table renders only with single-line rows... not sure how to fix this, though...

Author: I tried to implement this but without success:
https://rackerlabs.github.io/docs-rackspace/tools/rtd-tables.html

All steps were already done, except setting the html_context variable, which did not work.

What do you think @aperezhortal ?

Reviewer 2: Sorry for the delay. I missed this issue :S 

Commit 72fe8b5 should fix this :)  This hack also fixes the problems with the of the single-line rows in the Developers documentation.

   :widths: 30 20 20 20
   :header-rows: 1

   * - Nowcast type
     - Machine learning
     - Pysteps
     - Verification
   * - Deterministic (variance-preserving)
     - NA?
     - pysteps.nowcasts.extrapolation (test different optical flow methods)
     - MSE, RMSE, MAE, ETS, etc
   * - Deterministic (error-minimization)
     - ANNs, (deep) CNNs, random forests, AdaBoost, support vector regression, etc
     - pysteps.nowcasts.sprog, pysteps.nowcasts.anvil or ensemble mean of pysteps.nowcasts.steps/sseps
     - MSE, RMSE, MAE, ETS, etc or better normalized scores, etc
   * - Probabilistic (quantile-based)
     - Quantile ANN, quantile random forests, quantile regression
     - Probabilities derived from pysteps.nowcasts.steps/sseps
     - Reliability diagram (predicted vs observed quantile), probability integral transform (PIT) histogram
   * - Probabilistic (ensemble-based)

>      - GANNs, VAEs, etc

Reviewer: ```suggestion
     - GANs, VAEs, etc
```
